source/simulate/profile-simulation.md

```diff
@@ -0,0 +1,497 @@
+
+
+@functools.lru_cache(maxsize = 1000)
+def instant_3(instant: Instant) -> Instant:
+    return periods.Instant(instant)
+```
+
+Let's profile again:
+
+```
+Line #  Hits    Time        Per Hit    % Time         Line Contents
+===================================================================
+
+355       50    1787843.0   35756.9     98.8          parameters_at_instant = self.parameters.get_at_instant(str(instant))
+16     44100      92290.0       2.1     14.1          instant = str(periods.instant(instant))
+48        50        349.0       7.0     22.2          integers = [parse_fragment(fragment) for fragment in fragments]
+213       17         76.0       4.5    100.0          return int(fragment)
+```
+
+And overall?
+
+```
+PYTEST_ADDOPTS="$PYTEST_ADDOPTS --durations=3" openfisca test --country-package openfisca_france tests/formulas/irpp_prets_participatifs.yaml
+
+...
+
+7.33s call     tests/formulas/irpp_prets_participatifs.yaml::
+7.16s call     tests/formulas/irpp_prets_participatifs.yaml::
+2.46s call     tests/formulas/irpp_prets_participatifs.yaml::
+```
+
+Looks promising!
+
+## Beware of context!
+
+So we tried reducing the number of hits, what about reducing the impact per hit?
+
+Looking back to `get_at_instant`, let's follow the next function call:
+
+```python
+Line #  Hits     Time      Per Hit    % Time         Line Contents
+===================================================================
+
+16     44100      76860.0      1.7     15.4          instant = str(periods.instant(instant))
+17     44100     421350.0      9.6     84.6          return self._get_at_instant(instant)
+```
+
+Let's look at that code closer:
+
+```python
+def _get_at_instant(self, instant):
+    for value_at_instant in self.values_list:
+        if value_at_instant.instant_str <= instant:
+            return value_at_instant.value
+    return None
+```
+
+A linear search! I know what you're thinking, what if we do a binary search instead?
+
+Let's try that out (naive implementation):
+
+```python
+def _get_at_instant(self, instant):
+    index = self.values_dict.bisect_right(instant)
+
+    if index > len(self.values_list):
+        return None
+
+    return self.values_list[::-1][index - 1].value
+```
+
+And performance wise?
+
+```
+Line #  Hits     Time      Per Hit    % Time         Line Contents
+===================================================================
+
+16     44100      82468.0      1.9     10.6          instant = str(periods.instant(instant))
+17     44100     696812.0     15.8     89.4          return self._get_at_instant(instant)
+177   196150     596129.0      3.0     65.8          index = self.values_dict.bisect_right(instant)
+```
+
+It's actually worse... but why?
+
+One hypothesis: even if, compared to a linear search with complexity _O(n)_, a binary search should more efficient in that it has a complexity of _O(log(n))_, it will actually be more efficient for large numbers of _n_, which is not usually the case here.
```

Comment:
```suggestion
One hypothesis is, even if, compared to a linear search with complexity _O(n)_, a binary search should be more efficient in that it has a complexity of _O(log(n))_, it will actually be more efficient for large numbers of _n_, which is not usually the case here.
```